download script

crdanielbusch opened the pull request:
## Description

Write a script that downloads all climate data from the FAOSTAT website.
We need to:
* create a `downloaded_data` directory
* download the zip files
* unpack the zip files

## Checklist

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Changelog item added to `changelog/`

## Notes
* In the script it is best to call datalad download-url via datalad api, then the location of the data is saved and it is automatically unlocked. otherwise you would always have to unlock the data before the script runs
* FAOSTAT has no version numbers in the data and quasi rolling releases, which makes it a bit more difficult to see when new data is actually available. But maybe there is something in the data files
* the individual domains also have bulk downloads, that would also be an option. then we could even download the metadata as well, then we have it right along with the data
* zip files can be pushed to gin as well
* We need to use selenium to load javascript and load the html text. Only beautiful soup and requests doesn't work
* Data is now saved in a folder named after the lates update, e.g. `2023-11-09`
* Updated `.gitattributes` to consider `.zip` files as well

JGuetschow commented on scripts/remove_downloads.py:
```diff
@@ -0,0 +1,42 @@
+from faostat_data_primap.helper.definitions import downloaded_data_path
+
+
+def run():
```
What is the delete function good for? Why would we want to delete version?

crdanielbusch replied:
We don't really need this script. I've used it to test the work flow. Once I have the tests set up I can delete it

JGuetschow commented on scripts/remove_downloads.py:
```diff
@@ -0,0 +1,42 @@
+            files_to_delete = os.listdir(path_to_files)
+
+            for file in files_to_delete:
+                path_to_file = path_to_files / file
```
Deletion of committed files only works after unlocking them using datalad unlock

crdanielbusch replied:
Good to know

JGuetschow commented on src/faostat_data_primap/download.py:
```diff
@@ -0,0 +1,149 @@
+
+    Returns
+    -------
+        True if the file was downloaded, False if a cached file was found
```
In case there is no file, but download fails, requests throws and error, correct?

crdanielbusch replied:
Pretty sure it does. I can try and make a test for that

JGuetschow commented on scripts/download_all_domains.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def download_all_domains(sources: list[tuple[str]]) -> list[str]:
+    """
```
You could add downloading of the metadata available with the data. Having the methodology descriptions next to the data could be quite helpful

JGuetschow replied:
But It's less important and can be added later.

JGuetschow replied:
I would be good to keep old version of the methodology description docs though, to see if things have changed. I assume it will not always be updated, so we might want to use checksums and only store new versions of they have changed and symlink to the old version instead. That could quickly rule out methodology changes (if they are not updated yearly)

JGuetschow reviewed (approved):
Looks good and ready to go ahead with data reading. I've added a few minor comments.

JGuetschow reviewed (approved):
Looks good and ready to go ahead with data reading. I've added a few minor comments.

crdanielbusch commented:
@JGuetschow this is ready to be reviewed again.

Since you've looked at it last time:
* I have added the possibility to download the methodology document - that's the `download_methodology` function. 
* Added more tests - unit tests for download functions and one integration test that downloads all available data.
* I had to tweak some CI / pre-commit checks because I wasn't able to make them work: dis-abled the `poetry-check` command in the pre-commit (got a weird network error, we can also have a look at it together), and had to de-activate poetry cache in the github action (this is a known issue, see https://gitlab.com/climate-resource/copier-core-python-repository/-/issues/38)

JGuetschow commented on src/faostat_data_primap/download.py:
```diff
@@ -1,20 +1,162 @@
+    # Attempt to find a file to compare in the previous release
+    if previous_release:
+        file_to_compare = previous_release / filename
+        if file_to_compare.exists():
```
If a file can not be downloaded in the 30s period the script will fail, correct? I thin this is fine for now, but might need changes if the FAO servers are not always fast with responses (using multiple tries).

JGuetschow commented on src/faostat_data_primap/download.py:
```diff
@@ -31,7 +173,9 @@ def get_html_content(url: str) -> bs4.BeautifulSoup:
     # will automatically download it for you. Make sure there is no
```
Why make sure there is no chrome driver installed? Typo?

crdanielbusch replied:
Sometimes Selenium asks you to update your Chrome driver to the latest version. With the latest version of Selenium you don't need to do this anymore and Selenium will download the correct driver for you. However, that only works if you do not have a Chrome driver version installed on your system. I got an error that said I have to delete my Chrome driver. But, I can see how this comment may be more confusing than helpful, so I think it's better to delete it.

JGuetschow reviewed (approved):
Looks all good. I've not checked the tests code in detail, but from skimming it I think it all makes sense.
